services/resource-usage-tracker/src/simcore_service_resource_usage_tracker/modules/db/repositories/resource_tracker.py

```diff
@@ -1,131 +1,267 @@
+                        == data.service_run_id
+                    )
+                    & (
+                        resource_tracker_service_runs.c.service_run_status
+                        == ServiceRunStatus.RUNNING
+                    )
+                    & (
+                        resource_tracker_service_runs.c.last_heartbeat_at
+                        <= data.last_heartbeat_at
+                    )
+                )
+                .returning(resource_tracker_service_runs.c.service_run_id)
             )
+            result = await conn.execute(update_stmt)
+        row = result.first()
+        if row is None:
+            return None
+        return row[0]
 
-            on_update_stmt = insert_stmt.on_conflict_do_update(
-                index_elements=[
-                    resource_tracker_container.c.container_id,
-                ],
-                set_={
-                    "container_cpu_usage_seconds_total": sa.func.greatest(
-                        resource_tracker_container.c.container_cpu_usage_seconds_total,
-                        insert_stmt.excluded.container_cpu_usage_seconds_total,
-                    ),
-                    "prometheus_created": sa.func.least(
-                        resource_tracker_container.c.prometheus_created,
-                        insert_stmt.excluded.prometheus_created,
-                    ),
-                    "prometheus_last_scraped": sa.func.greatest(
-                        resource_tracker_container.c.prometheus_last_scraped,
-                        insert_stmt.excluded.prometheus_last_scraped,
-                    ),
-                    "modified": sa.func.now(),
-                },
+    async def update_service_run_stopped_at(
+        self, data: ServiceRunStoppedAtUpdate
+    ) -> ServiceRunId | None:
+        async with self.db_engine.begin() as conn:
+            update_stmt = (
+                resource_tracker_service_runs.update()
+                .values(
+                    modified=sa.func.now(),
+                    stopped_at=data.stopped_at,
+                    service_run_status=data.service_run_status,
+                )
+                .where(
+                    (
+                        resource_tracker_service_runs.c.service_run_id
+                        == data.service_run_id
+                    )
+                    & (
+                        resource_tracker_service_runs.c.service_run_status  ## .is_
+                        == ServiceRunStatus.RUNNING
+                    )
+                )
+                .returning(resource_tracker_service_runs.c.service_run_id)
             )
+            result = await conn.execute(update_stmt)
+        row = result.first()
+        if row is None:
+            return None
+        return row[0]
 
-            await conn.execute(on_update_stmt)
-
-    async def list_containers_by_user_and_product(
+    async def list_service_runs_by_user_and_product(
```

Comment:
Q: how many times will this be called? you have no indexes on the fields over which you are making queries here. This will be slow, with the ever growing number of entries.

Reply:
Good catch thanks, I will add indexes.